backend/howtheyvote/api/query.py

```diff
@@ -168,80 +188,92 @@ def where(self, expression: ColumnElement[Any]) -> Self:
         return query
 
 
-class MeilisearchSearchParams(TypedDict):
-    limit: int
-    offset: int
-    attributesToRetrieve: list[str]
-    filter: list[str]
-    sort: list[str]
+class ValueDecayWeightPostingSource(ValuePostingSource):
+    """A posting source that returns a weight that decreases the larger the difference between
+    a given constant origin and a document value is. If the document value equals the given
+    origin, the posting source returns a weight of 1. The weight decreases linearly until the
+    maximum difference is reached, at which point the posting source returns a weight of 0.
+
+    For example, this can be used to boost recent documents by using the current time as the
+    origin.
+
+    In order to adjust the range of the returned weight, combine this with a OP_SCALE_WEIGHT
+    query. For example, to assign a weight between 0 and 0.5, use a scaling factor of 0.5.
+
+    See: https://getting-started-with-xapian.readthedocs.io/en/latest/advanced/postingsource.html
+    """
+
+    def set_max_diff(self, max_diff: float) -> None:
+        self.max_diff = max_diff
+
+    def set_origin(self, origin: float) -> None:
+        self.origin = origin
+
+    def get_weight(self) -> float:
```

Comment:
Not sure how efficient this is, as I assume there’s a performance overhead due to Xapian having to call a Python method for every matching document during ranking. But I think at our scale this is pretty much irrelevant.